enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,477 @@
+
+```txt
+[Global]
+secret-name = openstack-credentials
+secret-namespace = kube-system
+```
+
+Resources deployed here will be destroyed with the bootstrap stage cleanup.
+
+#### Operator resource management strategy
+
+Every cloud manages it's configuration differently, but shares a common overall structure.
+
+1. `cloud-controller-manager` and related resources are deployed under a single namespace.
+2. Each pod (`CCM`, `cloud-node-manager`, etc.) will be managed by a `Deployment`.
+3. The pods are expected to run on the control-plane `Nodes` using `hostNetwork`.
+
+Operator will manage:
+
+- Dynamic provider resources, such as `Deployments`.
+- Operator will populate, create and manage the cloud-specific set of `Deployments`, based on the selected `platform` in the `Infrastructure` resource.
+- `Deployments`, provisioned by the operator will run the pods on the `control-plane` nodes, and will tolerate `cloudprovider.kubernetes.io/uninitialized` and `node-role.kubernetes.io/master` taints. They will run at least `2` replicas of the `CCM` pods with anti-affinity settings to prevent scheduling on the same machine and allow some fault-tolerance.
+- The `--cloud-provider` and the `--cloud-config` arguments will be populated using existing `library-go` based implementation from the [cluster-kube-controller-manager-operator](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/990ea3f2ace13be6579c00a889842c3f5d3e756a/pkg/operator/configobservation/configobservercontroller/observe_config_controller.go#L82-L85) to share a common approach with the bootstrap specific static-pod configuration.
+- Operator will own a `cluster-operator` resource, which will report the readiness of all workloads at post-install phase. The conditions reported will help other dependent components, such as [openshift-ingress](https://github.com/openshift/cluster-ingress-operator).
+- Operator will be built with consideration for potential future implementations of hybrid OpenShift clusters which support multiple cloud providers simultaneously (ie ensure resources created for each provider by the operator do not clash with resources for other providers)
+
+*Note: static resource creation is delegated to CVO (SA, RBAC, etc.)* 
+
+#### CVO managemnet
+
+The operator image will be build and included in every release payload, and expected to be deployed and running at kubernetes operators level ([10-29](https://github.com/openshift/cluster-version-operator/blob/34010292c3abd2582b687e9c0ef76c5924998f39/docs/dev/operators.md#how-do-i-get-added-as-a-special-run-level)) right after the `kube-controller-manager-operator` at runlevel [25](https://github.com/openshift/cluster-kube-controller-manager-operator/tree/master/manifests).
+
+Proposed `CCMO` manifests runlevel is `26`.
+
+#### Cloud provider management
+
+Each cloud provider will be hosted under its own repository. Those repositories will be initially bootstrapped by forking upstream ones.
```

Comment:
Why forking? What changes are needed that you need to fork from the start? I'd imagine us using upstream wherever possible.

Reply:
Is it not standard in Red Hat to fork a project into one of our orgs so that we can set up our own CI and build/release automation? Like we do for the cluster-autoscaler for example.

Is there some other way we can build the artifacts required without forking?

Reply:
We will try to be but in order to preserve openshift requirements on security, we might want to use a different set of default ports, services, containers for each provider.

Reply:
We also need to support a couple of releases down the line, have a custom Dockerfile and a well-known QA coverage for the CCM code, before putting it into the release.

Reply:
> We will try to be but in order to preserve openshift requirements on security, we might want to use a different set of default ports, services, containers for each provider.

This is configuration done by the operator, it should not require fork.

Dockerfile and RH built-related stuff are reasonable, though.